### ISSUE TITLE
alter('insert_col') and alter('insert_row') don't support null as an index

### ISSUE THREAD

**Issue by YakovL** [handsontable/handsontable#4448]
The issue is described [at StackOverflow](https://stackoverflow.com/questions/45405852/handsontable-adding-a-column-moves-colheaders-how-to-fix) (a snippet is attached there): when in a `beforeKeyDown` hook `this.alter('insert_col');` is used and `colHeaders` are used, adding a new column on the right moves the headers. Using `this.alter('insert_col',endColNum);` [fixes the problem](http://jsfiddle.net/6kzjLoj7/), but [docs](https://docs.handsontable.com/0.33.0/Core.html#alter) say that it should work without the second argument for adding a column after the last one, so I'm considering this as a bug.

Sorry, haven't made a minimal example yet, but the unimportant bits can be easily removed (the `if(event.which == $left && !isFirstCol)` block etc) so it should be something like this:

    var example = document.getElementById('example1'),
        hot = new Handsontable(example,{
      data: Handsontable.helper.createSpreadsheetData(2, 3),
      colHeaders: ["one","two","three"],
      contextMenu: true
    });

    Handsontable.hooks.add('beforeKeyDown',function(event)
    {
        var $right = 39,
            selected = this.getSelected(),
            isEditMode = this.getActiveEditor().isOpened();
        if(isEditMode) return;

        // calc dimensions
        var endColNum   = selected ? (selected[3]+1) : null,
            colsNum = this.countCols(),
            isLastCol   = endColNum == colsNum;

        // handle the right key
        if(isLastCol && event.which == $right)
            this.alter('insert_col');
    });

**Comment by AMBudnik** [comment]
Have you tried something like this http://jsfiddle.net/vv16be8c/?
I am not sure if it meets your project requirements - let me know if it does.

**Comment by YakovL** [comment]
Well, like I said, the workaround is already found, I'm just pointing that the docs suggest that the second argument in `.alter('insert_col',endColNum)` can be omitted ("If index is null or undefined, the new column
will be added after the last column."), which sounds like "this is the same as `.alter('insert_col')`", but the latter is not in fact correct. I'd consider this as a bug.

**Comment by AMBudnik** [comment]
You're right, the `alter` method should not shift the headers in this way. The easiest example is at 

http://jsfiddle.net/4x56q55u/

![xcgif](https://user-images.githubusercontent.com/10757813/29016983-0a816f00-7b55-11e7-951d-56b77dda5a11.gif)

we should get `[One, Two, A]`

**Comment by melloc01** [comment]
Any news on this issue?

**Comment by AMBudnik** [comment]
Sorry @melloc01 but the issue hasn't been raised by a developer yet

**Comment by melloc01** [comment]
@AMBudnik it's cool, thanks

For those who still suffering from this, I solved it by just **DISABLING** the colheaders and making my own headers with React, it seems a lot of work but it isn't - and definitely worth it =)

**Comment by AMBudnik** [comment]
Thanks for sharing the workaround. I would have never thought of that.

**Comment by melloc01** [comment]
Also put some selects on the headers.

![image](https://user-images.githubusercontent.com/4994258/29728002-8fb74dd8-89ac-11e7-83c6-8a9a05fe3522.png)

**Comment by YakovL** [comment]
@melloc01 I think some guys who find this thread because they have this problem too will be thankful to you if you share code examples (not just screenshots)

**Comment by melloc01** [comment]
Will do ASAP @YakovL, but just to clarify, it is just some react code reacting to handsontable [callbacks](https://docs.handsontable.com/0.34.0/tutorial-using-callbacks.html)

**Comment by AMBudnik** [comment]
Works the same with `8.0.0-rc1 build 26/09/2019`

http://jsfiddle.net/handsoncode/f0rw5x9k/

**Comment by AMBudnik** [comment]
Works the same in 7.4.2 http://jsfiddle.net/f602spne/ 8.0.0-beta1 http://jsfiddle.net/qnfv15s7/ and latest 8.0.0-beta2-rev19 http://jsfiddle.net/qnfv15s7/1/

**Comment by sl01k** [comment]
This has been resolved in Handsontable v17. The ambiguous alter('insert_col') API has been replaced with explicit alter('insert_col_start') and alter('insert_col_end'), which eliminates the null index confusion. Column headers are no longer shifted when inserting columns.

Demo: https://jsfiddle.net/nmoqxbh7/1/

Closing as resolved.
